daemon/execdriver/windows/namedpipes.go

```diff
@@ -6,77 +6,63 @@ import (
 	"io"
 
 	"github.com/Sirupsen/logrus"
-	"github.com/natefinch/npipe"
+	"github.com/docker/docker/daemon/execdriver"
 )
 
-// stdinAccept runs as a go function. It waits for the container system
-// to accept our offer of a named pipe for stdin. Once accepted, if we are
-// running "attached" to the container (eg docker run -i), then we spin up
-// a further thread to copy anything from the client into the container.
-//
-// Important design note. This function is run as a go function for a very
-// good reason. The named pipe Accept call is blocking until one of two things
-// happen. Either someone connects to it, or it is forcibly closed. Let's
-// assume that no-one connects to it, the only way otherwise the Run()
-// method would continue is by closing it. However, as that would be the same
-// thread, it can't close it. Hence we run as another thread allowing Run()
-// to close the named pipe.
-func stdinAccept(inListen *npipe.PipeListener, pipeName string, copyfrom io.ReadCloser) {
+// General comment. Handling I/O for a container is very different to Linux.
+// We use a named pipe to HCS to copy I/O both in and out of the container,
+// very similar to how docker daemon communicates with a CLI.
 
-	// Wait for the pipe to be connected to by the shim
-	logrus.Debugln("stdinAccept: Waiting on ", pipeName)
-	stdinConn, err := inListen.Accept()
-	if err != nil {
-		logrus.Errorf("Failed to accept on pipe %s %s", pipeName, err)
-		return
-	}
-	logrus.Debugln("Connected to ", stdinConn.RemoteAddr())
+// startStdinCopy asynchronously copies an io.Reader to the container's
+// process's stdin pipe and closes the pipe when there is no more data to copy.
+func startStdinCopy(pipe io.WriteCloser, copyfrom io.Reader) {
 
 	// Anything that comes from the client stdin should be copied
 	// across to the stdin named pipe of the container.
 	if copyfrom != nil {
```

Comment:
Actually since the rework due to a comment from @tiborvass, given the only call to startStdin/outCopy is from setupPipes which already check for nil, there is no check needed. Removing.